### ISSUE TITLE
老师，为什么我的paddleNLP项目训练loss始终不下降？

### ISSUE THREAD

**Issue by qarzhy** [PaddlePaddle/PaddleNLP#2236]
```python
# 导入模块

import os
import time
import random
import numpy as np
import json
from sklearn.metrics import f1_score, accuracy_score
from paddlenlp.metrics.glue import AccuracyAndF1
from functools import partial

# 飞桨API
import paddle
import paddle.nn as nn
import paddle.fluid as fluid
import paddle.nn.functional as F
import paddlenlp as ppnlp
from paddlenlp.losses import RDropLoss
from paddlenlp.data import Stack, Tuple, Pad
from paddlenlp.datasets import load_dataset
from paddlenlp.transformers import LinearDecayWithWarmup

# 可视化
from visualdl import LogWriter
# 读取训练数据
def read_train_text(data_path):
    with open(data_path, 'r', encoding='utf-8') as f:
        for line in f:
            data = line.strip('\n')
            
            dup_len = random.randint(a=1, b=int(DUP_RATE*len(data)))
            dup_index = random.sample(list(range(len(data))), k=dup_len)

            dup_ls = []
            for idx, token in enumerate(data):
                if idx in dup_index:
                    dup_ls.append(token)
                dup_ls.append(token)

            dup_data = ''.join(dup_ls)

            yield {'text1': data, 'text2': dup_data}
# 读取验证数据
def read_test_text(data_path):
    with open(data_path, 'r', encoding='utf-8') as f:
        for line in f:
            text1, text2, label = eval(line)
            yield {'text1': text1, 'text2':text2, 'label':int(label)}

# 转换词向量
def embedding(data, tokenizer):
    result = []

    for key, text in data.items():
        if 'label' in key:
            result += [data['label']]
        else:
            encoded_inputs = tokenizer(text=text)
            input_ids = encoded_inputs['input_ids']
            token_type_ids = encoded_inputs['token_type_ids']
            result += [input_ids, token_type_ids]

    return result
# 数据管道
def create_dataloader(dataset, mode='train', batch_size=1, batchify_fn=None, trans_fn=None):
    if trans_fn:
        dataset = dataset.map(trans_fn)

    shuffle = True if mode == 'train' else False
    if mode == 'train':
        batch_sampler = paddle.io.DistributedBatchSampler(dataset, batch_size=batch_size, shuffle=shuffle)
    else:
        batch_sampler = paddle.io.BatchSampler(dataset, batch_size=batch_size, shuffle=shuffle)

    return paddle.io.DataLoader(dataset=dataset, batch_sampler=batch_sampler, collate_fn=batchify_fn, return_list=True)
class SEBT(nn.Layer):
    def __init__(self, ptm, bz, dr, output_emb_size):
        super().__init__()

        self.ptm = ptm
        self.bz = bz
        self.dr = nn.Dropout(dr if dr is not None else 0.1)
        self.emb = output_emb_size

        # # 维度变换，想得到output_emb_size维度的cls向量
        # if self.emb > 0:
        #     weight_attr = paddle.ParamAttr(initializer=nn.initializer.TruncatedNormal(std=0.02))
        #     self.linear_change_emb = paddle.nn.Linear(768, self.emb, weight_attr=weight_attr)

    def get_cls_embedding(self, input_ids, token_type_ids):
        
        _ , cls = self.ptm(input_ids, token_type_ids, None, None)

        # if self.emb > 0:
        #     cls = self.linear_change_emb(cls)
        
        cls = self.dr(cls)

        cls = F.normalize(cls, p=2, axis=-1)

        return cls


    def get_projected_embedding(self, input_ids, token_type_ids):
        
        cls = self.get_cls_embedding(input_ids, token_type_ids)

        project1 = fluid.layers.fc(input=cls, size = self.emb * 2, act='relu')
        project2 = fluid.layers.fc(input=project1, size=self.emb * 2, act='relu')
        project3 = fluid.layers.fc(input=project2, size=self.emb, act='relu')

        # 批次归一化
        batch_norm = nn.BatchNorm1D(project3.shape[-1])
        project3 = batch_norm(project3)

        return project1



    def get_loader_cls(self, dataloader):
        self.eval()
        with paddle.no_grad():
            for batch_data in dataloader:
                input_ids1, token_type_ids1, input_ids2, token_type_ids2, label = batch_data

                input_ids1 = paddle.to_tensor(input_ids1)
                token_type_ids1 = paddle.to_tensor(token_type_ids1)
                input_ids2 = paddle.to_tensor(input_ids2)
                token_type_ids2 = paddle.to_tensor(token_type_ids2)

                cls1 = self.get_cls_embedding(input_ids1, token_type_ids1)
                cls2 = self.get_cls_embedding(input_ids2, token_type_ids2)

                yield [cls1, cls2, label]

    def forward(self, input_ids1, input_ids2, token_type_ids1, token_type_ids2, para):
        prj1 = self.get_projected_embedding(input_ids1, token_type_ids1)
        prj2 = self.get_projected_embedding(input_ids2, token_type_ids2)        
        cross_corr = prj1.T @ prj2
        cross_corr = cross_corr / self.bz

        identity_diag = paddle.full(shape=[prj1.shape[-1]], fill_value=1, dtype=paddle.get_default_dtype())

        same_dimension = paddle.sum((identity_diag - paddle.diag(cross_corr))**2)
        other_dimension = paddle.sum((cross_corr - paddle.diag(paddle.diag(cross_corr)))**2)
        
        loss = same_dimension + para * other_dimension
        
        return loss
def evaluate(model, dataloader):
    loader_cls = model.get_loader_cls(dataloader)
    sim_all = []
    label_all = []

    for idx, (cls1, cls2, label) in enumerate(loader_cls):
        sim_all += list(F.cosine_similarity(cls1, cls2).numpy())
        label_all += list(label.numpy())

    pred_all = np.array(sim_all) > 0.5
    pred_all = pred_all + 0

    f1 = f1_score(pred_all, np.array(label_all))
    acc = accuracy_score(pred_all, np.array(label_all))

    return f1, acc
def train():
    
    paddle.set_device(device)
    
    # load and process data
    
    train_ds = load_dataset(read_train_text, data_path='work/data/train.txt', lazy=False)
    dev_ds = load_dataset(read_test_text, data_path = 'work/data/test.txt', lazy=False)
    tokenizer = ppnlp.transformers.BertTokenizer.from_pretrained('bert-base-chinese')
     
    trans_func = partial(embedding, tokenizer=tokenizer)

    

    batchify_fn = lambda samples, fn=Tuple(
        Pad(axis=0, pad_val=tokenizer.pad_token_id),  # query_input
        Pad(axis=0, pad_val=tokenizer.pad_token_type_id),  # query_segment
        Pad(axis=0, pad_val=tokenizer.pad_token_id),  # title_input
        Pad(axis=0, pad_val=tokenizer.pad_token_type_id),  # tilte_segment
    ): [data for data in fn(samples)]

    batchify_fn_test = lambda samples, fn=Tuple(
        Pad(axis=0, pad_val=tokenizer.pad_token_id),  # query_input
        Pad(axis=0, pad_val=tokenizer.pad_token_type_id),  # query_segment
        Pad(axis=0, pad_val=tokenizer.pad_token_id),  # title_input
        Pad(axis=0, pad_val=tokenizer.pad_token_type_id),  # tilte_segment,
        Stack()
        ): [data for data in fn(samples)]

    train_loader = create_dataloader(train_ds, mode='train',batch_size=batch_size, batchify_fn=batchify_fn, trans_fn=trans_func)
    test_loader = create_dataloader(dev_ds, mode='test',batch_size=100, batchify_fn=batchify_fn_test, trans_fn=trans_func)
    for idx, batch in enumerate(test_loader, start=1):
        if idx == 1:
            test_batch = batch
    # configure model training
    ptm = ppnlp.transformers.BertModel.from_pretrained('bert-base-chinese')
    model = SEBT(ptm, bz=batch_size, dr=0.2, output_emb_size=output_emb_size)

    num_training_steps = len(train_loader) * num_epochs
    lr_scheduler = LinearDecayWithWarmup(
        learning_rate=learning_rate,
        total_steps=num_training_steps,
        warmup=warmup_proportion)
    decay_params = [
        p.name for n, p in model.named_parameters()
        if not any(nd in n for nd in ["bias", "norm"])
    ]
    grad_clip = paddle.nn.ClipGradByGlobalNorm(max_grad_norm)
    
    #sgd = paddle.optimizer.SGD(learning_rate=0.1, parameters=linear.parameters(), weight_decay=0.01)
    optimizer = paddle.optimizer.AdamW(
        learning_rate=lr_scheduler,
        parameters=model.parameters(),
        weight_decay=weight_decay,
        apply_decay_param_fun=lambda x: x in decay_params,
        grad_clip=grad_clip)

    metric = AccuracyAndF1()

    # start to train model
    global_step, best_f1 = 1, 0.
    model.train()
    for epoch in range(1, num_epochs + 1):
        for batch_data in train_loader():
            input_ids1, token_type_ids1, input_ids2, token_type_ids2 = batch_data
            loss = model(input_ids1=input_ids1, input_ids2=input_ids2, token_type_ids1=token_type_ids1, token_type_ids2=token_type_ids2, para=para)

            t_input_ids1, t_token_type_ids1, t_input_ids2, t_token_type_ids2, label = test_batch
            t_loss = model(input_ids1=t_input_ids1, input_ids2=t_input_ids2, token_type_ids1=t_token_type_ids1, token_type_ids2=t_token_type_ids2, para=para)
            
            # logits: batch_size, seql_len, num_tags
            # logits = model(input_ids, token_type_ids=token_type_ids)
            # loss = F.cross_entropy(logits, labels)

            loss.backward()
            lr_scheduler.step()
            optimizer.step()
            optimizer.clear_grad()

            if global_step > 0 and global_step % log_steps == 0:
                print(
                    f"epoch: {epoch} - global_step: {global_step}/{num_training_steps} - loss:{loss.numpy().item():.6f}"
                )
            if (global_step > 0 and global_step % eval_steps == 0
                ) or global_step == num_training_steps:
                f1,acc = evaluate(model, test_loader)
                model.train()
                if f1 > best_f1:
                    print(
                        f"best F1 performence has been updated: {best_f1:.5f} --> {f1:.5f}"
                    )
                    best_f1 = f1
                    paddle.save(model.state_dict(),
                                f"{checkpoints}/best.pdparams")
                print(
                    f'evalution result: accuracy:{acc:.5f} F1: {f1:.5f}'
                )

            global_step += 1

    paddle.save(model.state_dict(), f"{checkpoints}/final.pdparams")

log_steps = 50
eval_steps = 50
seed = 1000
device = 'gpu'
DUP_RATE = 0.8
num_epochs = 10
batch_size = 32
learning_rate = 0.5
max_seq_len = 512
output_emb_size = 768
para = 0.2
max_grad_norm = 1.0
warmup_proportion = 0.1
weight_decay = 0.01
checkpoints = 'work/'
train()
```
epoch: 1 - global_step: 50/940 - loss:1535.602539
best F1 performence has been updated: 0.00000 --> 0.73684
evalution result: accuracy:0.60000 F1: 0.73684
epoch: 2 - global_step: 100/940 - loss:1535.651611
best F1 performence has been updated: 0.73684 --> 0.74214
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 2 - global_step: 150/940 - loss:1535.656250
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 3 - global_step: 200/940 - loss:1535.666992
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 3 - global_step: 250/940 - loss:1535.658813
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 4 - global_step: 300/940 - loss:1535.628540
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 4 - global_step: 350/940 - loss:1535.636963
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 5 - global_step: 400/940 - loss:1535.636230
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 5 - global_step: 450/940 - loss:1535.656372
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 6 - global_step: 500/940 - loss:1535.615356
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 6 - global_step: 550/940 - loss:1535.621826
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 7 - global_step: 600/940 - loss:1535.636841
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 7 - global_step: 650/940 - loss:1535.616455
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 8 - global_step: 700/940 - loss:1535.619873
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 8 - global_step: 750/940 - loss:1535.619629
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 9 - global_step: 800/940 - loss:1535.633667
evalution result: accuracy:0.59000 F1: 0.74214
epoch: 10 - global_step: 850/940 - loss:153
[test.txt](https://github.com/PaddlePaddle/PaddleNLP/files/8722987/test.txt)

[train.txt](https://github.com/PaddlePaddle/PaddleNLP/files/8722985/train.txt)

**Comment by yingyibiao** [comment]
换一个小一点的学习率尝试一下，比如 1e-4
